### PR TITLE
Prevent attached models from disappearing during parent reload

### PR DESCRIPTION
**Pull request by Foghrye4**
Player models attached to e.g. boats can mysteriously disappear if they unload due
to zone-out and zone-in because child entities are not correctly sent to clients.

This fix updates the server to send attached entity data to the client explicitly for
those entities.

Fixing issue #3681 by adding child client initialisation data inside parent client initialisation data. After data received on client side, infants will be spawned if they are not exist or will be re-initialised if they already here.

### CODE REVIEW

File: src/content_cao.cpp [hunk 1]
```diff
@@ -1776,6 +1776,17 @@ void GenericCAO::processMessage(const std::string &data)
 			m_nametag->nametag_color = m_prop.nametag_color;
 		}
 	}
+	else if (cmd == GENERIC_CMD_SPAWN_INFANT) {
+		u16 child_id = readU16(is);
+		u8 type = readU8(is);
+
+		GenericCAO *childobj = m_env->getGenericCAO(child_id);
+		if (childobj == NULL) {
+			m_env->addActiveObject(child_id, type,	deSerializeLongString(is));
```
Comment (SmallJoker) [inline src/content_cao.cpp]: Tabulator found after `type,`

File: src/content_sao.cpp [hunk 2]
```diff
@@ -385,6 +385,12 @@ std::string LuaEntitySAO::getClientInitializationData(u16 protocol_version)
 			os<<serializeLongString(gob_cmd_update_bone_position((*ii).first, (*ii).second.X, (*ii).second.Y)); // m_bone_position.size
 		}
 		os<<serializeLongString(gob_cmd_update_attachment(m_attachment_parent_id, m_attachment_bone, m_attachment_position, m_attachment_rotation)); // 4
+		for(std::set<int>::const_iterator ii = m_attachment_child_ids.begin(); ii != m_attachment_child_ids.end(); ++ii){ // 7
```
Comment (SmallJoker) [inline src/content_sao.cpp]: Spaces after `if`, `for` etc.
Line is longer than >80 characters

**Comment by sofar** [comment]
The commit title shouldn't be "fix issue #foo", but something like "Fix disappearing player model when attached".

**Comment by Foghrye4** [comment]
Code style formatted, title changed.

**Comment by SmallJoker** [comment]
@Foghrye4 You forgot some spaces after `if`and `for` in your `code slyle` commit. Please review it.

**Comment by Foghrye4** [comment]
Frigging git.

**Comment by paramat** [comment]
Please can you squash commits?

**Comment by sofar** [comment]
lol @Foghrye4  - yes, git can be rough to handle.

I would suggest rephrasing the title and commit message. The title should be sort and descriptive. The commit message should explain the problem, and "fixes #xxx" should go near the bottom instead of the top.

Here's what I would have written (freely interpreted, may be wrong or omitting elements you wrote)

```
Prevent attached player models from disappearing during area reload

Player models attached to e.g. boats can mysteriously disappear if they unload due
to zone-out and zone-in because child entities are not correctly sent to clients.

This fix updates the server to send attached entity data to the client explicitly for
those entities.

Fixes #3681
```

**Comment by paramat** [comment]
Yes first line fairly short line to not wrap, a summary without a full stop, followed by an empty line, followed by details with full stops.

**Comment by Foghrye4** [comment]
@paramat i learned how to squash a day ago, using google. Didn't try it yet. I will try it later, as well it would be fine if you squash commits on merge.

**Comment by paramat** [comment]
The tutorials in the github help pages 'bootcamp' are the best way to learn, no hurry as learning git takes time, we can squash this.

**Comment by Foghrye4** [comment]
@paramat Ok, it seems I've done it.

**Comment by raymoo** [comment]
@Foghrye4 It's the commit title and message that needs to be changed, not the PR one.

**Comment by Foghrye4** [comment]
@raymoo yea, I missed that. Thank.

File: src/content_cao.cpp [hunk 3]
```diff
@@ -1776,6 +1776,17 @@ void GenericCAO::processMessage(const std::string &data)
 			m_nametag->nametag_color = m_prop.nametag_color;
 		}
 	}
+	else if (cmd == GENERIC_CMD_SPAWN_INFANT) {
```
Comment (nerzhul) [inline src/content_cao.cpp]: because now we have many commands, can you replace this if/elseif with switch case please ?

File: src/content_cao.cpp [hunk 4]
```diff
@@ -1776,6 +1776,17 @@ void GenericCAO::processMessage(const std::string &data)
...
+		u16 child_id = readU16(is);
+		u8 type = readU8(is);
+
+		GenericCAO *childobj = m_env->getGenericCAO(child_id);
```
Comment (nerzhul) [inline src/content_cao.cpp]: if prefer 

if (GenericCAO *childobj = m_env->getGenericCAO(child_id)) { 
do yes 
} else { 
do no 
}

File: src/content_sao.cpp [hunk 5]
```diff
@@ -385,6 +385,13 @@ std::string LuaEntitySAO::getClientInitializationData(u16 protocol_version)
 			os<<serializeLongString(gob_cmd_update_bone_position((*ii).first, (*ii).second.X, (*ii).second.Y)); // m_bone_position.size
 		}
 		os<<serializeLongString(gob_cmd_update_attachment(m_attachment_parent_id, m_attachment_bone, m_attachment_position, m_attachment_rotation)); // 4
+		for (std::set<int>::const_iterator ii = m_attachment_child_ids.begin(); 
+				ii != m_attachment_child_ids.end(); ++ii) {
+			ServerActiveObject *obj = m_env->getActiveObject(*ii);
+			if(obj){
```
Comment (nerzhul) [inline src/content_sao.cpp]: missing brace

File: src/content_sao.cpp [hunk 6]
```diff
@@ -385,6 +385,13 @@ std::string LuaEntitySAO::getClientInitializationData(u16 protocol_version)
 			os<<serializeLongString(gob_cmd_update_bone_position((*ii).first, (*ii).second.X, (*ii).second.Y)); // m_bone_position.size
 		}
 		os<<serializeLongString(gob_cmd_update_attachment(m_attachment_parent_id, m_attachment_bone, m_attachment_position, m_attachment_rotation)); // 4
+		for (std::set<int>::const_iterator ii = m_attachment_child_ids.begin();
```
Comment (nerzhul) [inline src/content_sao.cpp]: can you use UNORDERED_SET instead of std::set and change the container type of m_attachment_child_ids to this too ?

File: src/content_sao.cpp [hunk 7]
```diff
@@ -868,6 +875,13 @@ std::string PlayerSAO::getClientInitializationData(u16 protocol_version)
 				m_physics_override_jump, m_physics_override_gravity, m_physics_override_sneak,
 				m_physics_override_sneak_glitch)); // 5
 		os << serializeLongString(gob_cmd_update_nametag_attributes(m_prop.nametag_color)); // 6 (GENERIC_CMD_UPDATE_NAMETAG_ATTRIBUTES) : Deprecated, for backwards compatibility only.
+		for (std::set<int>::const_iterator ii = m_attachment_child_ids.begin(); 
+				ii != m_attachment_child_ids.end(); ++ii) {
+			ServerActiveObject *obj = m_env->getActiveObject(*ii);
+			if (obj) {
+				os<<serializeLongString(gob_cmd_update_infant(*ii, obj->getSendType(), obj->getClientInitializationData(protocol_version)));
```
Comment (nerzhul) [inline src/content_sao.cpp]: missing spaces

File: src/content_sao.cpp [hunk 8]
```diff
@@ -868,6 +875,13 @@ std::string PlayerSAO::getClientInitializationData(u16 protocol_version)
 				m_physics_override_jump, m_physics_override_gravity, m_physics_override_sneak,
 				m_physics_override_sneak_glitch)); // 5
 		os << serializeLongString(gob_cmd_update_nametag_attributes(m_prop.nametag_color)); // 6 (GENERIC_CMD_UPDATE_NAMETAG_ATTRIBUTES) : Deprecated, for backwards compatibility only.
+		for (std::set<int>::const_iterator ii = m_attachment_child_ids.begin(); 
+				ii != m_attachment_child_ids.end(); ++ii) {
+			ServerActiveObject *obj = m_env->getActiveObject(*ii);
```
Comment (nerzhul) [inline src/content_sao.cpp]: Push this to condition directly

File: src/content_sao.cpp [hunk 9]
```diff
@@ -385,6 +385,13 @@ std::string LuaEntitySAO::getClientInitializationData(u16 protocol_version)
 			os<<serializeLongString(gob_cmd_update_bone_position((*ii).first, (*ii).second.X, (*ii).second.Y)); // m_bone_position.size
 		}
 		os<<serializeLongString(gob_cmd_update_attachment(m_attachment_parent_id, m_attachment_bone, m_attachment_position, m_attachment_rotation)); // 4
+		for (std::set<int>::const_iterator ii = m_attachment_child_ids.begin(); 
+				ii != m_attachment_child_ids.end(); ++ii) {
+			ServerActiveObject *obj = m_env->getActiveObject(*ii);
```
Comment (nerzhul) [inline src/content_sao.cpp]: Push this to condition directly

File: src/content_sao.cpp [hunk 10]
```diff
@@ -868,6 +875,13 @@ std::string PlayerSAO::getClientInitializationData(u16 protocol_version)
 				m_physics_override_jump, m_physics_override_gravity, m_physics_override_sneak,
 				m_physics_override_sneak_glitch)); // 5
 		os << serializeLongString(gob_cmd_update_nametag_attributes(m_prop.nametag_color)); // 6 (GENERIC_CMD_UPDATE_NAMETAG_ATTRIBUTES) : Deprecated, for backwards compatibility only.
+		for (std::set<int>::const_iterator ii = m_attachment_child_ids.begin();
```
Comment (nerzhul) [inline src/content_sao.cpp]: same here, UNORDERED_SET

File: src/genericobject.cpp [hunk 11]
```diff
@@ -182,3 +182,15 @@ std::string gob_cmd_update_nametag_attributes(video::SColor color)
 	writeARGB8(os, color);
 	return os.str();
 }
+
+std::string gob_cmd_update_infant(u16 id, u8 type, std::string client_initialization_data)
```
Comment (nerzhul) [inline src/genericobject.cpp]: change this proto to : 
void gob_cmd_update_infant(const u16 id, const u8 type, const std::string &client_initialization_data, std::string &result)

and write os.str() to result  instead of return os.str()

**Comment by nerzhul** [comment]
@Foghrye4 you didn't need to squash since recent github changes, we can do it ourselves, but yo ucan do it yourself manually to have proper integration

**Comment by Megaf** [comment]
@Foghrye4 Will this also fix the issue that player names are not following the model if they are in a boat? This is other related issue we have.

**Comment by Foghrye4** [comment]
@nerzhul roger that. In a process.

**Comment by Foghrye4** [comment]
@Megaf probably no.

File: src/content_sao.cpp (hunk 5 above)
Comment (Foghrye4) [inline src/content_sao.cpp]: I don't get it. "{"?

File: src/content_sao.cpp (hunk 7 above)
Comment (Foghrye4) [inline src/content_sao.cpp]: Do you mean 'os << serializeLongString(gob_cmd_update_infant(*ii, obj->getSendType(), obj->getClientInitializationData(protocol_version)));'?

File: src/genericobject.cpp (hunk 11 above)
Comment (Foghrye4) [inline src/genericobject.cpp]: @nerzhul I can't do so. Because at content_sao side its:
'os<<serializeLongString(gob_cmd_update_infant(...'
Therefore strings passed is serialized two times. One time only _init. data_, second time - _init. data_ with _id_ and _type_.

Comment (Foghrye4) [inline src/genericobject.cpp]: @nerzhul on a second through, I will change all old functions a same way.

File: src/content_cao.cpp [hunk 12]
```diff
@@ -1572,220 +1572,216 @@ void GenericCAO::processMessage(const std::string &data)
...
+			// Place us a bit higher if we're physical, to not sink into
+			// the ground due to sucky collision detection...
+			if (m_prop.physical)
+				m_position += v3f(0,0.002,0);
+
+			if (getParent()) // Just in case
+				return;
+
+			if (do_interpolate)
 			{
```
Comment (nerzhul) [inline src/content_cao.cpp]: This brace should be on upper line

File: src/content_cao.cpp [hunk 13]
```diff
@@ -1572,220 +1572,216 @@ void GenericCAO::processMessage(const std::string &data)
...
-
-		GenericCAO *childobj = m_env->getGenericCAO(child_id);
-		if (childobj == NULL) {
-			m_env->addActiveObject(child_id, type, deSerializeLongString(is));
-		} else {
-			childobj->initialize(deSerializeLongString(is));
-		}
+			break;
+		default:
+			warningstream<<FUNCTION_NAME
```
Comment (nerzhul) [inline src/content_cao.cpp]: spaces " << "

File: src/genericobject.h [hunk 14]
```diff
@@ -81,4 +81,3 @@ std::string gob_cmd_update_nametag_attributes(video::SColor color);
 std::string gob_cmd_update_infant(u16 id, u8 type, std::string client_initialization_data);
 
 #endif
-
```
Comment (nerzhul) [inline src/genericobject.h]: You should keep this empty line

File: src/genericobject.cpp (hunk 11 above)
Comment (Zeno-) [inline src/genericobject.cpp]: I think this should remain as it is, so that it remains consistent with all the other functions that genericobject.cpp provides.

File: src/content_cao.cpp (hunk 3 above)
Comment (Zeno-) [inline src/content_cao.cpp]: @nerzhul I think that should be done in a separate commit.

File: src/content_cao.cpp [hunk 15]
```diff
@@ -1572,209 +1572,216 @@ void GenericCAO::processMessage(const std::string &data)
...
-			m_animation_range = v2s32((s32)range.X, (s32)range.Y);
-			m_animation_speed = readF1000(is);
-			m_animation_blend = readF1000(is);
-			// these are sent inverted so we get true when the server sends nothing
-			m_animation_loop = !readU8(is);
-			updateAnimation();
-		} else {
-			LocalPlayer *player = m_env->getLocalPlayer();
-			if(player->last_animation == NO_ANIM)
+	switch (cmd) {
```
Comment (SmallJoker) [inline src/content_cao.cpp]: Notice: It is allowed to remove a whole level of indentation after `switch (cmd) {`
https://www.kernel.org/doc/Documentation/CodingStyle (Chapter 1: Indentation)

File: src/content_cao.cpp [hunk 16]
```diff
@@ -1572,209 +1572,216 @@ void GenericCAO::processMessage(const std::string &data)
...
+					// As there is no definition, make a smoke puff
+					ClientSimpleObject *simple = createSmokePuff(
+							m_smgr, m_env, m_position,
+							m_prop.visual_size * BS);
+					m_env->addSimpleObject(simple);
+				} else {
+					// TODO: Execute defined fast response
+					// Flashing shall suffice as there is no definition
+					m_reset_textures_timer = 0.05;
+					if(damage >= 2)
```
Comment (SmallJoker) [inline src/content_cao.cpp]: Space after `if`

**Comment by nerzhul** [comment]
Can you also rebase, src/content_sao.cpp is now in conflict ?

Thanks

**Comment by Foghrye4** [comment]
@nerzhul done.
